flink-clients/src/main/java/org/apache/flink/client/FlinkYarnSessionCli.java

```diff
@@ -0,0 +1,429 @@
+
+		// Jar Path
+		Path localJarPath;
+		if(cmd.hasOption(FLINK_JAR.getOpt())) {
+			String userPath = cmd.getOptionValue(FLINK_JAR.getOpt());
+			if(!userPath.startsWith("file://")) {
+				userPath = "file://" + userPath;
+			}
+			localJarPath = new Path(userPath);
+		} else {
+			LOG.info("No path for the flink jar passed. Using the location of "+flinkYarnClient.getClass()+" to locate the jar");
+			localJarPath = new Path("file://"+flinkYarnClient.getClass().getProtectionDomain().getCodeSource().getLocation().getPath());
+			if(!localJarPath.toString().contains("uberjar")) {
+				// we need to have a proper uberjar because otherwise we don't have the required classes available on the cluster.
+				// most likely the user did try to start yarn in a regular hadoop2 flink build (not a yarn package) (using ./bin/flink -m yarn-cluster)
+				LOG.error("The detected jar file '"+localJarPath+"' is not a uberjar.");
+				return null;
+			}
+		}
+
+		flinkYarnClient.setLocalJarPath(localJarPath);
+
+		// Conf Path
+		String confDirPath = CliFrontend.getConfigurationDirectoryFromEnv();
+		GlobalConfiguration.loadConfiguration(confDirPath);
+		flinkYarnClient.setConfigurationDirectory(confDirPath);
+		File confFile = new File(confDirPath + File.separator + CONFIG_FILE_NAME);
+		if(!confFile.exists()) {
+			LOG.error("Unable to locate configuration file in "+confFile);
+			return null;
+		}
+		Path confPath = new Path(confFile.getAbsolutePath());
+
+		flinkYarnClient.setConfigurationFilePath(confPath);
+
+		List<File> shipFiles = new ArrayList<File>();
+		// path to directory to ship
+		if(cmd.hasOption(SHIP_PATH.getOpt())) {
+			String shipPath = cmd.getOptionValue(SHIP_PATH.getOpt());
+			File shipDir = new File(shipPath);
+			if(shipDir.isDirectory()) {
+				shipFiles = new ArrayList<File>(Arrays.asList(shipDir.listFiles(new FilenameFilter() {
+					@Override
+					public boolean accept(File dir, String name) {
+						return !(name.equals(".") || name.equals(".."));
+					}
+				})));
+			} else {
+				LOG.warn("Ship directory is not a directory. Ignoring it.");
+			}
+		}
+
+		//check if there is a logback or log4j file
+		if(confDirPath.length() > 0) {
+			File logback = new File(confDirPath + File.pathSeparator + CONFIG_FILE_LOGBACK_NAME);
+			if(logback.exists()) {
+				shipFiles.add(logback);
+				flinkYarnClient.setConfigurationFilePath(new Path(logback.toURI()));
+			}
+			File log4j = new File(confDirPath + File.pathSeparator + CONFIG_FILE_LOG4J_NAME);
+			if(log4j.exists()) {
+				shipFiles.add(log4j);
+				if(flinkYarnClient.getFlinkLoggingConfigurationPath() != null) {
+					// this means there is already a logback configuration file --> fail
+					LOG.error("The configuration directory ('"+confDirPath+"') contains both LOG4J and Logback configuration files." +
+							"Please delete or rename one of them.");
+					return null;
+				} // else
+				flinkYarnClient.setConfigurationFilePath(new Path(log4j.toURI()));
+			}
+		}
```

Comment:
Can we check that the corresponding logging backend will actually be used by Flink? Because the system should rather pick the config file whose backend is present.